File: src/main/java/io/codehive/bloc4j/lib/ShaderProgram.java
package io.codehive.bloc4j.lib;

import static com.jogamp.opengl.GL.GL_FALSE;
import static com.jogamp.opengl.GL2ES2.GL_COMPILE_STATUS;
import static com.jogamp.opengl.GL2ES2.GL_FRAGMENT_SHADER;
import static com.jogamp.opengl.GL2ES2.GL_INFO_LOG_LENGTH;
import static com.jogamp.opengl.GL2ES2.GL_LINK_STATUS;
import static com.jogamp.opengl.GL2ES2.GL_VERTEX_SHADER;

import com.jogamp.opengl.GL3;
import com.jogamp.opengl.util.GLBuffers;
import io.codehive.bloc4j.Main;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.nio.charset.StandardCharsets;
import org.apache.commons.io.IOUtils;

public class ShaderProgram {

  private final GL3 gl;
  private final String vertexShaderFile;
  private final String fragmentShaderFile;

  private int shaderProgram;
  private boolean built = false;

  public ShaderProgram(GL3 gl, String vertexShaderFile, String fragmentShaderFile) {
    this.gl = gl;
    this.vertexShaderFile = vertexShaderFile;
    this.fragmentShaderFile = fragmentShaderFile;
  }

  public void build() {
    int vertexShader = createShader(GL_VERTEX_SHADER, vertexShaderFile);
    int fragmentShader = createShader(GL_FRAGMENT_SHADER, fragmentShaderFile);

    shaderProgram = createProgram(vertexShader, fragmentShader);
    built = true;
  }

  public void use() {
    if (!built) {
      throw new IllegalStateException("Attempted using Shader Program before building it");
    }
    gl.glUseProgram(shaderProgram);
  }

  public void delete() {
    gl.glDeleteProgram(shaderProgram);
  }

  private int createProgram(int vertexShader, int fragmentShader) {
    int program = gl.glCreateProgram();
    gl.glAttachShader(program, vertexShader);
    gl.glAttachShader(program, fragmentShader);
    gl.glLinkProgram(program);

    IntBuffer success = GLBuffers.newDirectIntBuffer(1);
    gl.glGetProgramiv(program, GL_LINK_STATUS, success);
    if (success.get(0) == GL_FALSE) {
      IntBuffer infoLogLength = GLBuffers.newDirectIntBuffer(1);
      gl.glGetProgramiv(program, GL_INFO_LOG_LENGTH, infoLogLength);

      ByteBuffer bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength.get(0));
      gl.glGetProgramInfoLog(program, infoLogLength.get(0), null, bufferInfoLog);
      byte[] bytes = new byte[infoLogLength.get(0)];
      bufferInfoLog.get(bytes);
      String strInfoLog = new String(bytes);
      System.err.printf("Error linking Shader Program: %s%n", strInfoLog);
    }
    gl.glDeleteShader(vertexShader);
    gl.glDeleteShader(fragmentShader);
    return program;
  }

  private int createShader(int shaderType, String fileName) {
    int shader = gl.glCreateShader(shaderType);
    String shaderSource = getShaderSourceCode(fileName);
    IntBuffer length = GLBuffers.newDirectIntBuffer(new int[]{shaderSource.length()});
    gl.glShaderSource(shader, 1, new String[]{shaderSource}, length);
    gl.glCompileShader(shader);

    IntBuffer success = GLBuffers.newDirectIntBuffer(1);
    gl.glGetShaderiv(shader, GL_COMPILE_STATUS, success);
    if (success.get(0) == GL_FALSE) {
      IntBuffer infoLogLength = GLBuffers.newDirectIntBuffer(1);
      gl.glGetShaderiv(shader, GL_INFO_LOG_LENGTH, infoLogLength);

      ByteBuffer bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength.get(0));
      gl.glGetShaderInfoLog(shader, infoLogLength.get(0), null, bufferInfoLog);
      byte[] bytes = new byte[infoLogLength.get(0)];
      bufferInfoLog.get(bytes);
      String strInfoLog = new String(bytes);
      System.err.printf("Error compiling %s Shader: %s%n", shaderTypeName(shaderType), strInfoLog);
    }
    return shader;
  }

  private String getShaderSourceCode(String fileName) {
    try {
      return IOUtils.toString(Main.class.getResourceAsStream(fileName), StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private String shaderTypeName(int shaderType) {
    return switch (shaderType) {
      case GL_VERTEX_SHADER -> "Vertex";
      case GL_FRAGMENT_SHADER -> "Fragment";
      default -> "";
    };
  }
}

File: src/main/kotlin/io/codehive/bloc4j/lib/ShaderProgram.kt
package io.codehive.bloc4j.lib

import com.jogamp.opengl.GL
import com.jogamp.opengl.GL2ES2.*
import com.jogamp.opengl.GL3
import com.jogamp.opengl.util.GLBuffers
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.nio.charset.StandardCharsets

class ShaderProgram(
  private val gl: GL3,
  private val vertexShaderFile: String,
  private val fragmentShaderFile: String
) {
  private var shaderProgram = 0
  private var built = false

  fun build() {
    val vertexShader = createShader(GL_VERTEX_SHADER, vertexShaderFile)
    val fragmentShader = createShader(
      GL_FRAGMENT_SHADER, fragmentShaderFile
    )

    shaderProgram = createProgram(vertexShader, fragmentShader)
    built = true
  }

  fun use() {
    check(built) { "Attempted using Shader Program before building it" }
    gl.glUseProgram(shaderProgram)
  }

  fun delete() {
    gl.glDeleteProgram(shaderProgram)
  }

  private fun createProgram(vertexShader: Int, fragmentShader: Int): Int {
    val program = gl.glCreateProgram()
    gl.glAttachShader(program, vertexShader)
    gl.glAttachShader(program, fragmentShader)
    gl.glLinkProgram(program)

    val success = GLBuffers.newDirectIntBuffer(1)
    gl.glGetProgramiv(program, GL_LINK_STATUS, success)
    if (success[0] == GL.GL_FALSE) {
      val infoLogLength = GLBuffers.newDirectIntBuffer(1)
      gl.glGetProgramiv(program, GL_INFO_LOG_LENGTH, infoLogLength)

      val bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength[0])
      gl.glGetProgramInfoLog(program, infoLogLength[0], null, bufferInfoLog)
      val bytes = ByteArray(infoLogLength[0])
      bufferInfoLog.get(bytes)
      val strInfoLog = String(bytes)
      System.err.println("Error linking Shader Program: $strInfoLog")
    }
    gl.glDeleteShader(vertexShader)
    gl.glDeleteShader(fragmentShader)
    return program
  }

  private fun createShader(shaderType: Int, fileName: String): Int {
    val shader = gl.glCreateShader(shaderType)
    val shaderSource = getShaderSourceCode(fileName)
    val length = GLBuffers.newDirectIntBuffer(intArrayOf(shaderSource.length))
    gl.glShaderSource(shader, 1, arrayOf(shaderSource), length)
    gl.glCompileShader(shader)

    val success = GLBuffers.newDirectIntBuffer(1)
    gl.glGetShaderiv(shader, GL_COMPILE_STATUS, success)
    if (success[0] == GL.GL_FALSE) {
      val infoLogLength = GLBuffers.newDirectIntBuffer(1)
      gl.glGetShaderiv(shader, GL_INFO_LOG_LENGTH, infoLogLength)

      val bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength[0])
      gl.glGetShaderInfoLog(shader, infoLogLength[0], null, bufferInfoLog)
      val bytes = ByteArray(infoLogLength[0])
      bufferInfoLog[bytes]
      val strInfoLog = String(bytes)
      System.err.println("Error compiling ${shaderTypeName(shaderType)} Shader: $strInfoLog")
    }
    return shader
  }

  private fun getShaderSourceCode(fileName: String): String {
    return try {
      IOUtils.toString(
        javaClass.getResourceAsStream(fileName),
        StandardCharsets.UTF_8
      )
    } catch (e: IOException) {
      throw RuntimeException(e)
    }
  }

  private fun shaderTypeName(shaderType: Int): String {
    return when (shaderType) {
      GL_VERTEX_SHADER -> "Vertex"
      GL_FRAGMENT_SHADER -> "Fragment"
      else -> ""
    }
  }
}
